sudo/src/pam.rs

```diff
@@ -71,37 +74,119 @@ fn determine_auth_status(
     }
 }
 
-pub fn authenticate(context: &Context) -> Result<(), Error> {
-    let authenticate_for = &context.current_user.name;
-    let target_user = context.target_user.uid;
-
-    // init pam
-    let mut pam = sudo_pam::PamContext::builder_cli(context.stdin)
-        .target_user(authenticate_for)
-        .service_name("sudo")
-        .build()?;
-    pam.mark_silent(true);
-    pam.mark_allow_null_auth_token(false);
-
-    // determine session limit
-    let scope = determine_record_scope(context);
-
-    // only if there is an interactive terminal or parent process we can store session information
-    let (must_authenticate, records_file) = determine_auth_status(scope, context);
-
-    if must_authenticate {
-        pam.authenticate()?;
-        if let (Some(mut session_records), Some(scope)) = (records_file, scope) {
-            match session_records.create(scope, target_user) {
-                Ok(_) => (),
-                Err(e) => {
-                    auth_warn!("Could not update session record file with new record: {e}");
+type PamBuilder<C> = dyn Fn(&Context) -> PamResult<PamContext<C>>;
```

Comment:
One concern I have with this making things very abstract here: this is code that (at some point) should really be very clear that we are authenticating properly (also since we are really relying on the integration tests for this part, and they have limitations). By making the `PamBuilder` something that can in principle be anything there really needs to be a clear spot somewhere where it's made clear that this is not a "smoke screen".

Reply:
(Of course, by making the fields with this type in `PamAuthenticator` non-public, there is no 'security gap' here)